### PR TITLE
Remove indirection layer in at_index_impl

### PR DESCRIPTION
**Pull request by msimberg**
Retry of #4975 since I merged #4975 too early.

### CODE REVIEW

**Review by hkaiser — DISMISSED** [review]
LGTM, thanks!

**Comment by K-ballo** [comment]
That's a nice warning, this implementation is wrong.. was it always wrong?

**Comment by msimberg** [comment]
> That's a nice warning, this implementation is wrong.. was it always wrong?

Feel free to ignore that last commit completely if necessary. It's most likely just me who just made it wrong...

**Comment by K-ballo** [comment]
> It's most likely just me who just made it wrong...

No, it was always wrong... The empty overload was never even considered.

The original implementation had a selection function that wasn't templated on the index (see http://talesofcpp.fusionfenix.com/post-22/true-story-efficient-packing#the-case-of-the-%60tuple%60), that's why those were members of the implementation. For some reason, when I added this to HPX I chose to add a seemingly redundant template parameter `J` (I must have had _some_ reason), but failed to add one to the fallback overload.

The proper fix is to add `template <std::size_t J>` to the empty overload.

File: libs/core/type_support/include/hpx/type_support/pack.hpp [hunk 1]
```diff
@@ -177,22 +173,23 @@ namespace hpx { namespace util {
         {
         };
 
+        template <std::size_t J>
+        static empty at_index_check(...);
```
Comment (msimberg) [inline libs/core/type_support/include/hpx/type_support/pack.hpp:177]: @K-ballo this is what you meant?

Comment (K-ballo) [reply libs/core/type_support/include/hpx/type_support/pack.hpp:177]: Yes, now this overload will be included in the `at_index_check<I>` call.

**Pull request merged**